# TRATAMENTO DE EXCE√á√ïES
    ü°∫ excess√£o √© um evento que ocorre durante a a excecu√ß√£o do programa, que interro,pe p fluxo normal de execu√ß√£o
    ü°∫
    ü°∫

## Exemplos de erro 

**exemplo de erro 1 - tentando acessar um elemento fora do indice de um array**
```.java
package exemplo1;
public class Ex1 {
	public static void main(String[] args) {
		int []  numeros = {1,2,3} ;
		System.out.println(numeros[3]); //ERRO ESPERADO : ArrayIndexOutOfBoundsException 
	}
}
```

**exemplo de erro 2 - divis√£o por zero**

```.java
package exemplo2;
public class Ex2 {
	public static void main(String[] args) {	
		int a = 10, b=0;
		System.out.println(a/b); //ERRO ESPERADO : ArithmeticException
	}
}
```

**exemplo de erro 3 - tentando converter uma string em um n√∫mero**
```.java
package exemplo2;
public class Ex3 {
	public static void main(String[] args) {
		String numero = "abc";
		int valor = Integer.parseInt(numero); // ERRO ESPERADO : NumberFormatException
	}
}

```
## TRATANDO EXCE√á√ïES
	ü°∫
 	ü°∫
**Tratando erro 1**
```.java
package exemplo1;

public class Ex1 {

	public static void main(String[] args) {
		int []  numeros = {1,2,3};
		try { //TRATADA O ERRO COMO UMA EXCE√á√ÉO
		System.out.println(numeros[3]);
		} catch(Exception e) {
			System.out.println("Exce√ß√£o:" + e.getMessage()); //FORMAS DE MOSTRAR QUAL A EXE√á√ÉO - APENAS MOSTRAR O QUE ESTA ERRADO
			System.out.println("Exce√ß√£o:" + e.toString()); // MOSTRA O QUE ESTA ERRADO E QUAL √â O ERRO
		}
		
		System.out.println("o programa segue em excecu√ß√£o");
		System.out.println(numeros[3]); //FORA DO TRY, AQUI VAI GERAR UM ERRO FATAL 
		System.out.println("agora n√£o"); // AQUI J√Å N√ÉO √â MAIS EXECUTADO O PROGRAMA


	}

}

```
```.java
package exemplo1;

public class Ex1 {

	public static void main(String[] args) {
		int []  numeros = {1,2,3};
		try { //TRATADA O ERRO COMO UMA EXCE√á√ÉO
		System.out.println(numeros[2]);
		System.err.println("resultado"+(3/0));
		} catch(ArrayIndexOutOfBoundsException e) { // SE USAR DESSA MANEIRA ELE TRATA APENAS O ERRO ESPECIFICO - se ocorrer algum outro erro ele n√£o √© tratado 
			System.out.println("Exce√ß√£o:" + e.getMessage()); //FORMAS DE MOSTRAR QUAL A EXE√á√ÉO
			System.out.println("Exce√ß√£o:" + e.toString());
		}catch(Exception e ) { // TRATA O ERRO DE FORMA MAIS GENERICA
			System.out.println("Exce√ß√£o:" + e.toString());

		}
		
		System.out.println("o programa segue em excecu√ß√£o");
		System.out.println(numeros[3]); //FORA DO TRY, AQUI VAI GERAR UM ERRO FATAL 
		System.out.println("agora n√£o"); // AQUI J√Å N√ÉO √â MAIS EXECUTADO O PROGRAMA


	}

}

```

**tratando erro 2**
```.java
package exemplo2;
public class Ex2 {
	public static void main(String[] args) {
		int a = 10, b=0;
		try {
		System.out.println(a/b); //ERRO ESPERADO : ArithmeticException
		}catch(ArithmeticException e) {
			System.out.println("Exce√ß√£o: "+e.toString());
		}
		System.out.println("CONTINUA");
	}
}

```
**tratando erro 3**
```.java
package exemplo2;
public class Ex3 {
	public static void main(String[] args) {
		String numero = "abc";
		try {
		int valor = Integer.parseInt(numero); // ERRO ESPERADO : NumberFormatException
		}catch(NumberFormatException e){
			System.out.println("exce√ß√£o: "+e.toString() );
		}
		System.out.println("CONTINUA");
	}
}

```

